pennylane/operation.py

```diff
@@ -331,9 +337,19 @@ def __init__(self, *params, wires=None, do_queue=True):
             self.queue()
 
     def __str__(self):
-        """Print the operator name and some information."""
+        """Operator name and some information."""
         return  "{}: {} params, wires {}".format(self.name, len(self.params), self.wires)
 
+    def __repr__(self):
+        """Constructor-call-like representation."""
+        # FIXME using self.parameters here instead of self.params is dangerous, it assumes the params can be evaluated
+        # which is only true if something suitable happens to remain in VariableRef.positional_arg_values etc. after
+        # the last evaluation.
```

Comment:
Is there a valid user pattern where this is not the case? Could add an xfail test for it.

Reply:
It seems quite unsafe, say you have two qnodes, evaluate both, and then call `print_applied` on both. The method will try to evaluate the parameters of both qnodes, but only the second one has its parameter values stored in `VariableRef.positional_arg_values`. The first one will get the wrong values, or may even request a value with an index that's too high.

That's the inherent downside of using global variables, you cannot assume they will stay valid between function calls.

Reply:
I can add a test but we should think of a way of fixing this somehow.

Reply:
Yep, I agree this should be fixed. This could also help fix some race conditions we are seeing with parallelization.

Reply:
One solution would be to remove the class variables `VariableRef.positional_arg_values` etc. and instead let each QNode keep their own "last evaluated parameters" instance variables. The `VariableRef` instances would have to have references to them so they can find them, but that should be no problem. This is more or less how Strawberry fields does this. What do you think?

Reply:
I think that sounds good - are there any downsides to changing to this approach? Sounds like this should be an ADR first.

Reply:
I can make an ADR and see if I can think of downsides. Maybe one more pointer dereference per evaluation?